File: BOJ/seulah/8_week/상어 초등학교.py
n = int(input())
students = [list(map(int, input().split())) for _ in range(n**2)]

data = [[0] * n for _ in range(n)]
dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

for student in students:
    max_prefer, max_empty = -1,-1
    x,y = 0,0
    for i in range(n):
        for j in range(n):
            # 빈자리가 있다면
            if data[i][j] == 0:
                prefer, empty = 0, 0

                # 상하좌우 확인
                for k in range(4):
                    nx = i + dx[k]
                    ny = j + dy[k]

                    if 0 <= nx < n and 0 <= ny < n:
                        # 좋아하는 학생이 주위에 있다면
                        if data[nx][ny] in student[1:]:
                            prefer += 1
                        # 빈자리가 있다면
                        if data[nx][ny] == 0:
                            empty += 1

                # 조건 1) 좋아하는 학생이 가장 많이 인접해있는 곳
                if prefer > max_prefer:
                    max_prefer = prefer
                    max_empty = empty  # 빈자리 개수도 갱신
                    x, y = i, j  # 최적의 자리 갱신

                # 조건 2) 1을 만족하는 가장 많은 인접해있는 곳도 여러 개면
                elif max_prefer == prefer:
                    if empty > max_empty:
                        max_empty = empty  # 빈자리가 가장 많은 곳 선택
                        x, y = i, j  # 최적의 자리 갱신
                    # 조건 3) 2를 만족하는 가장 많은 빈자리도 여러 개면
                    elif max_empty == empty:
                        if i < x:  # 행이 더 작은 곳을 선택
                            x , y = i, j
                        elif i == x and j < y:  # 행이 같다면 열이 더 작은 곳을 선택
                            x, y = i, j

    data[x][y] = student[0]  # 최적의 자리에 학생 배치

# 만족도 계산
satisfaction = 0
score = {0: 0, 1: 1, 2: 10, 3: 100, 4: 1000}
students.sort() #학생 번호를 기준으로 정렬

for i in range(n):
    for j in range(n):
        student_num = data[i][j]
        prefer = 0
        for k in range(4):
            nx = dx[k]+i
            ny = dy[k]+j
            if 0 <= nx < n and 0 <= ny < n:
                # 좋아하는 학생이 주위에 있다면
                if data[nx][ny] in students[student_num-1][1:]:
                    prefer += 1
        satisfaction += score[prefer]  # 최종 만족도

print(satisfaction)

File: BOJ/jaehong/8_week/상어 초등학교.js
const fs = require("fs");
const input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

const N = parseInt(input[0]);
const students = input.slice(1).map((line) => line.split(" ").map(Number));

console.log(sharkElementarySchool(N, students));

function sharkElementarySchool(N, students) {
  // 1. 교실 초기화
  let classroom = Array.from({ length: N }, () => Array(N).fill(0));

  // 2. 학생별 좋아하는 학생 저장
  let studentMap = new Map();
  students.forEach(([num, ...likes]) => {
    studentMap.set(num, new Set(likes));
  });

  // 3. 이동 방향 (상, 하, 좌, 우)
  const dx = [-1, 1, 0, 0];
  const dy = [0, 0, -1, 1];

  // 4. 자리 배치
  students.forEach(([num, ...likes]) => {
    let bestPositions = [];
    let maxLike = -1; // 좋아하는 학생 인접 수 최대값
    let maxEmpty = -1; // 빈 칸 개수 최대값

    for (let r = 0; r < N; r++) {
      for (let c = 0; c < N; c++) {
        if (classroom[r][c] !== 0) continue; // 이미 자리 차있음

        let likeCount = 0;
        let emptyCount = 0;

        for (let i = 0; i < 4; i++) {
          let nr = r + dx[i];
          let nc = c + dy[i];

          if (nr >= 0 && nr < N && nc >= 0 && nc < N) {
            if (studentMap.get(num).has(classroom[nr][nc])) {
              likeCount++;
            } else if (classroom[nr][nc] === 0) {
              emptyCount++;
            }
          }
        }

        // 조건 우선순위 결정
        if (likeCount > maxLike || (likeCount === maxLike && emptyCount > maxEmpty)) {
          bestPositions = [[r, c]];
          maxLike = likeCount;
          maxEmpty = emptyCount;
        } else if (likeCount === maxLike && emptyCount === maxEmpty) {
          bestPositions.push([r, c]);
        }
      }
    }

    // 3번 조건 (행/열 기준 정렬)
    bestPositions.sort((a, b) => a[0] - b[0] || a[1] - b[1]);
    let [selectedR, selectedC] = bestPositions[0];
    classroom[selectedR][selectedC] = num;
  });

  // 5. 만족도 계산
  let totalSatisfaction = 0;
  const satisfactionMap = [0, 1, 10, 100, 1000];

  for (let r = 0; r < N; r++) {
    for (let c = 0; c < N; c++) {
      let studentNum = classroom[r][c];
      let likes = studentMap.get(studentNum);

      let likeCount = 0;
      for (let i = 0; i < 4; i++) {
        let nr = r + dx[i];
        let nc = c + dy[i];

        if (nr >= 0 && nr < N && nc >= 0 && nc < N) {
          if (likes.has(classroom[nr][nc])) {
            likeCount++;
          }
        }
      }

      totalSatisfaction += satisfactionMap[likeCount];
    }
  }

  return totalSatisfaction;
}
